docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,164 @@
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+We will continue to use these config files unchanged, and custom files can be specified with an absolute path. Relative paths are assumed to be relative to the Python package e.g. `configs/skills/...`.
+
+### Future - Model Serving
+
+Custom pipelines may have more unique model serving requirements. Instead of serving just one model, we may need to launch the model server with a model and an additional model with adapter. vLLM, for example, can host both a model and a model+adapter under two different model IDs.
+
+The pipeline author needs some way of disambiguating between these multiple models - i.e. the definition of each `LLMBlock` needs to specify a particular model.
+
+Right now the `Pipeline` constructor takes two relevant parameters - the OpenAI client instance, and the model ID for the default model. It's important to note that this model ID is defined by the user at runtime, and it may not match the model IDs that the pipeline author used.
+
+The use cases will be:
+
+1. Most LLMBlock definitions will use the default teacher model - and we can make the semantic that if the pipeline author doesn't specify a model in an `LLMBlock`, the default in `PipelineContext.model_id` is used.
+2. In cases where a model+adapter is served, it should usually be the pipeline author defining both the model serving configuration and the model's use in an `LLMBlock`. So the pipeline author can name the model in the serving config and reference it by that name in the pipeline config.
+3. In cases where a model+adapter is served with a name that doesn't match that chosen by the pipeline author, the user can supply a mapping between the name used in the `LLMBlock` and the name used in the user's serving config.
+
+In future, it may make sense to define some model serving config (e.g. which adapter to use for SDG) as part of a flow definition. We will address this in a future if enhancement, if needed.
+
+### CLI Integration
+
+As of the current version of `ilab`, it supports `simple` and `full` as parameters to `--pipeline` to select one of the two types of built-in pipelines included in the library. Once we have support for loading custom pipelines, we need a way for these to be specified with the CLI. Here is a proposed CLI UX for this:
+
+> `ilab data generate`
+
+Use the default pipeline, `simple`.
+
+> `ilab data generate --pipeline full`
+
+Use the built-in `full` pipeline.
+
+> `ilab data generate --pipeline full --pipeline-extend path/to/config.yaml`
```

Comment:
again to repeat my previous comment, I don't think it makes sense to have a way to chain workflows like this and also this is very much out of the scope of this design PR.

Reply:
Some good discussion on that [here](https://github.com/instructlab/dev-docs/pull/109#discussion_r1668463920)

Somehow I didn't update the doc with the latest proposal there - chaining would be done by specifying `--pipeline` multiple times

But, personally, I'd be happy to omit that feature for now until there's a strong use case for it

Reply:
This all is intended to support the downstream to desire to ship custom pipeline content that would run after “full”. If we don’t expose extending in some way, we would ship a copy of “full” plus more from somewhere else and have to try to keep them in sync when fixes or improvements come up in one place or the other. 

Both work. The chaining thing lets us define stuff once. Indeed it adds some complexity to the CLI interface, though. 

Either way we have to be clear on how we intend to deal with that use case. Happy to hear other alternatives!

Reply:
> This all is intended to support the downstream to desire to ship custom pipeline content that would run after “full”.

Wow, I did not realize that - I thought the downstream pipeline was standalone, no dependence on an upstream pipeline. My bad. I'll capture that in the doc. (Unless Kai corrects us on this point, of course)

Reply:
We could make it that way. But the input to the custom parts is the output of “full” … so it’s this or we copy stuff or something else 🤷‍♂️

Reply:
Ok, I hope commit 7f24fdc clarifies this

I would like an ack from @xukai92 on:

> We believe the most common case for custom pipelines is for them to extend the `full` pipeline

Reply:
> We could make it that way. But the input to the custom parts is the output of “full” … so it’s this or we copy stuff or something else 🤷‍♂️

I don't think "copy stuff" is acceptable (I know you don't either) - every change to the full pipeline would need to be replicated to any custom pipeline

Reply:
> I would like an ack from @xukai92 on:
> 
> > We believe the most common case for custom pipelines is for them to extend the `full` pipeline

I asked @shivchander about this on Slack and got a confirmation. However, it's slightly more complicated than I thought!

In one case, we want to run custom configuration *after* the `full` pipeline (skills). In another, we want to run custom configuration *before* the `full` pipeline (knowledge). I missed this important ordering detail on the knowledge side.

We've got to factor this into the design somehow.

Reply:
trying to reply to some comments from @markmc 

> I would like an ack from @xukai92 on:
> We believe the most common case for custom pipelines is for them to extend the full pipeline

I would say "reusing blocks" is the more expected common use cases, not extending the (full) pipeline.
Ideally the flexibility we want is to reuse the blocks as nodes in a DAG.
It happens to that the extension from 1.5 to 2.0 is like appending a list of blocks to another list of blocks (which looks like we are chaining two pipelines. preview: putting existing + some new blocks in a DAG with some loops is what we will work on for something like 3.0).
But I don't think we need to nail this generic design down here as it's not a GA block.

> We could make it that way. But the input to the custom parts is the output of “full” … so it’s this or we copy stuff or something else 🤷‍♂️
> I don't think "copy stuff" is acceptable (I know you don't either) - every change to the full pipeline would need to be replicated to any custom pipeline

Please see mine and Russel's comment above.

And my feeling to copying stuff here is pretty mixed as the reusability/customization at the pipeline level is sort of limited.

Reply:
> In one case, we want to run custom configuration _after_ the `full` pipeline (skills). In another, we want to run custom configuration _before_ the `full` pipeline (knowledge). I missed this important ordering detail on the knowledge side.

Awesome to get this clarified :+1:

Reply:
>But I don't think we need to nail this generic design down here as it's not a GA block.

Great. In this kind of scenario, I'm looking out for ways our decisions now might prevent us adding support for more elaborate chaining like that in future. And it feels like some that can be added in a future release when we have a better understanding exactly what we need.

> And my feeling to copying stuff here is pretty mixed as the reusability/customization at the pipeline level is sort of limited.

"Copying stuff" is always an option of course. But if "chain two pipelines" together works for our first release, then supporting that will make the whole thing will be much more maintainable if we don't copy entire chains of blocks.

Reply:
@markmc had some good ideas on how we could expand `ImportBlock` to load specific blocks from another config instead of only allowing importing an entire config. It sounds like we have a pretty good path to add that in the near future if it makes sense.